src/app/utils/util-types.ts

```diff
@@ -0,0 +1,27 @@
+/** Generic TypeScript type utilities */
+
+/** Extracts the type of elements of an array */
+export type ArrayElement<ArrayType extends readonly unknown[]> =
+  ArrayType extends readonly (infer ElementType)[] ? ElementType : never;
+
+/**
+ * A lookup table from key to value, where not all keys may be mapped. We use these often
+ * to special case some logic for certain subsets of strings or hashes. Use Record<K,V> if
+ * the table is meant to be complete for all possible values of K. This can also be helpful
+ * to re-type imported JSON files as lookup tables.
+ */
+export type LookupTable<K extends keyof any, V> = {
+  readonly [P in K]?: V | undefined;
+};
+
+/**
+ * A convenience for a lookup table keyed by a hash (number). This also accepts strings
+ * since you can use string version of numbers to read into objects keyed by number just fine,
+ * and JSON files are always keyed by string.
+ */
+export type HashLookup<V> = LookupTable<number | string, V>;
+
+/**
+ * A convenience for a lookup table keyed by a string. Equivalent to NodeJS.ReadonlyDict.
+ */
+export type StringLookup<V> = LookupTable<string, V>;
```

Comment:
```suggestion
export type StringLookup<V> = LookupTable<string, V>;

/** performs `key in obj` but properly narrows `key` */
export function isIn<O extends Record<string | number, any>>(key: keyof O, obj: O): key is keyof O {
  return key in obj;
}

```

Reply:
check THIS shit out

Reply:
Nice! I'd done something similar in another project. The one tough part is that `Object.keys` really does stringily the keys, so this is relying explicitly on JS auto converting strings to numbers.

Reply:
that's fair yeah a lot of the upstream types are all wrong and fake and claim to be keyed by number. oh well.

Reply:
With an inline annotation terser will probably remove it. I also wish typescript did more with `in` (and other things it should theoretically be able to narrow from, like `Boolean`, but since it's technically possible to have a property be defined and set to `undefined` it can't really do much for you:

```
> const foo = { bar: undefined }
undefined
> foo['bar']
undefined
> 'bar' in foo
true
```